File: ss-won/programmers/60058.js
const isEmpty = (p) => {
  if (p.length == 0) return true;
  else false;
};

const dividing = (p) => {
  let u,
    v = "";
  let parr = p.split("");
  let finished = 0;
  let i = 0;
  do {
    if (parr[i] == "(") finished--;
    else finished++;
    i++;
  } while (finished != 0);

  u = parr.slice(0, i).join("");
  v = parr.slice(i, parr.length).join("");
  return [u, v];
};

const isRight = (u) => {
  let res = 0;
  let size = u.length;
  let tmp = u.split("");
  for (let next of tmp) {
    if (next == "(") res += size;
    else res -= size;
    size--;
  }
  return res;
};

const solution = (p) => {
  let answer = "";
  let u,
    v = "";

  /*1단계*/
  if (isEmpty(p)) return answer;
  /*2단계*/
  let res = dividing(p);
  u = res[0];
  v = res[1];
  /*3단계*/
  if (isRight(u) > 0) {
    //올바른 문자열
    u += solution(v);
    answer += u;
  } else {
    //올바른 문자열이 아님
    let tmp = "";
    tmp += "(";
    tmp += solution(v);
    tmp += ")";
    u = u.slice(1, u.length - 1);
    tmp += u
      .split("")
      .map((v) => {
        if (v == "(") return ")";
        else return "(";
      })
      .join("");
    answer += tmp;
  }
  return answer;
};

File: ss-won/programmers/60058.py
def isValidBracket(s):
    op, cs = 0, 0
    for b in s:
        if cs > op:
            return False
        if b == '(':
            op += 1
        else:
            cs += 1
    if op == cs:
        return True
    return False


def isBalancedBracket(s):
    op, cs = 0, 0
    for b in s:
        if b == '(':
            op += 1
        else:
            cs += 1
    if op == cs:
        return True
    return False


def Reverse(s):
    res = ""
    for i in range(len(s)):
        if s[i] == '(':
            res += ')'
        else:
            res += '('
    return res


def solution(p):
    answer, u, v = '', '', ''
    if len(p) == 0:
        return answer

    i = 2
    while i <= len(p):
        if isBalancedBracket(p[0:i]):
            u, v = p[0:i], p[i:]
            break
        i += 2

    if isValidBracket(u) == True:
        answer = u + solution(v)
        return answer

    answer = "(" + solution(v) + ")" + Reverse(u[1:len(u)-1])
    return answer
